File: movie_translator/subtitles/processor.py
"""Unified subtitle processor combining parsing, writing, and validation."""

from collections.abc import Callable
from pathlib import Path

from ..logging import logger
from ..types import NON_DIALOGUE_STYLES, DialogueLine, replace_polish_chars
from ._pysubs2 import get_pysubs2


class SubtitleProcessingError(Exception):
    """Base exception for subtitle processing errors."""

    pass


def _find_dialogue_style(subs) -> str:
    """Find the best dialogue style name in an SSAFile.

    Picks a style that exists in the file and is likely dialogue.
    Handles ASS files (with named styles like 'Dialogue', 'Default'),
    SRT-sourced files (always 'Default'), and edge cases where no
    styles are defined.
    """
    if not subs.styles:
        return 'Default'

    # Prefer 'Default' if it exists (SRT, most Polish subs, simple ASS)
    if 'Default' in subs.styles:
        return 'Default'

    # Look for common dialogue style names (case-sensitive match)
    dialogue_names = ('Dialogue', 'Dialog', 'Main', 'Dialogi', 'Normal')
    for name in dialogue_names:
        if name in subs.styles:
            return name

    # Case-insensitive search for anything with 'dialog' or 'default'
    for name in subs.styles:
        lower = name.lower()
        if 'dialog' in lower or 'default' in lower or 'main' in lower:
            return name

    # Last resort: use the first style and hope for the best
    return next(iter(subs.styles))


class SubtitleProcessor:
    """Unified subtitle processor for parsing, writing, and validation."""

    @staticmethod
    def extract_dialogue_lines(subtitle_file: Path) -> list[DialogueLine]:
        """Extract dialogue lines from subtitle file."""
        logger.info(f'📖 Reading {subtitle_file.name}...')

        if not subtitle_file.exists():
            raise SubtitleProcessingError(f'Subtitle file not found: {subtitle_file}')

        pysubs2 = get_pysubs2()
        if pysubs2 is None:
            raise SubtitleProcessingError('pysubs2 library not available')

        try:
            subs = pysubs2.load(str(subtitle_file))
        except Exception as e:
            raise SubtitleProcessingError(f'Failed to parse subtitle file: {e}') from e

        logger.debug(f'   - Loaded {len(subs)} total events')

        unique_subs = SubtitleProcessor._deduplicate_events(subs)
        dialogue_lines = SubtitleProcessor._filter_dialogue(unique_subs)

        return dialogue_lines

    @staticmethod
    def create_subtitle_file(
        original_file: Path,
        dialogue_lines: list[DialogueLine],
        output_path: Path,
        text_transform: Callable[[str], str] | None = None,
    ) -> None:
        """Create a new subtitle file from dialogue lines."""
        if not original_file.exists():
            raise SubtitleProcessingError(f'Original subtitle file not found: {original_file}')

        pysubs2 = get_pysubs2()
        if pysubs2 is None:
            raise SubtitleProcessingError('pysubs2 library not available')

        try:
            original_subs = pysubs2.load(str(original_file))
        except Exception as e:
            raise SubtitleProcessingError(f'Failed to load original subtitle file: {e}') from e

        new_subs = pysubs2.SSAFile()
        new_subs.info = original_subs.info.copy()
        new_subs.styles = original_subs.styles.copy()

        # Pick the dialogue style from the source file. The style must
        # exist in new_subs.styles or the player will use a bare fallback.
        dialogue_style = _find_dialogue_style(new_subs)

        for line in dialogue_lines:
            text = line.text
            if text_transform:
                text = text_transform(text)
            text = text.replace('\n', '\\N')

            event = pysubs2.SSAEvent(
                start=line.start_ms,
                end=line.end_ms,
                style=dialogue_style,
                text=text,
            )
            new_subs.append(event)

        try:
            new_subs.save(str(output_path))
        except Exception as e:
            raise SubtitleProcessingError(f'Failed to save subtitle file: {e}') from e

        logger.debug(f'   - Saved {len(new_subs)} events')

    @staticmethod
    def create_english_subtitles(
        original_file: Path,
        dialogue_lines: list[DialogueLine],
        output_path: Path,
    ) -> None:
        """Create clean English subtitle file with dialogue only."""
        logger.info(f'🔨 Creating clean English ASS: {output_path.name}')
        SubtitleProcessor.create_subtitle_file(original_file, dialogue_lines, output_path)
        logger.debug('   - Removed all non-dialogue events')

    @staticmethod
    def create_polish_subtitles(
        original_file: Path,
        translated_dialogue: list[DialogueLine],
        output_path: Path,
        replace_chars: bool = True,
    ) -> None:
        """Create Polish subtitle file with optional character replacement."""
        logger.info('🔤 Creating Polish subtitles')
        text_transform = replace_polish_chars if replace_chars else None
        SubtitleProcessor.create_subtitle_file(
            original_file, translated_dialogue, output_path, text_transform
        )

    @staticmethod
    def override_font_name(ass_file: Path, new_font_name: str) -> None:
        """Replace all font names in ASS styles with the given font name."""
        pysubs2 = get_pysubs2()
        if pysubs2 is None:
            raise SubtitleProcessingError('pysubs2 library not available')

        try:
            subs = pysubs2.load(str(ass_file))
        except Exception as e:
            raise SubtitleProcessingError(f'Failed to load subtitle file: {e}') from e

        for style in subs.styles.values():
            if hasattr(style, 'fontname'):
                style.fontname = new_font_name

        try:
            subs.save(str(ass_file))
        except Exception as e:
            raise SubtitleProcessingError(f'Failed to save subtitle file: {e}') from e

        logger.debug(f'   - Overrode font name to "{new_font_name}" in {ass_file.name}')

    @staticmethod
    def validate_cleaned_subtitles(original_file: Path, cleaned_file: Path) -> None:
        """Validate that cleaned subtitles maintain proper timing coverage."""
        pysubs2 = get_pysubs2()
        if pysubs2 is None:
            raise SubtitleProcessingError('pysubs2 library not available')

        try:
            original_subs = pysubs2.load(str(original_file))
            cleaned_subs = pysubs2.load(str(cleaned_file))
        except Exception as e:
            raise SubtitleProcessingError(f'Failed to load subtitle files: {e}') from e

        original_events = [e for e in original_subs if e.text.strip()]
        cleaned_events = [e for e in cleaned_subs if e.text.strip()]

        logger.debug(f'   📊 Validation: Original file has {len(original_events)} non-empty events')

        original_dialogue = [
            e
            for e in original_events
            if not any(kw in getattr(e, 'style', 'Default').lower() for kw in NON_DIALOGUE_STYLES)
        ]
        non_dialogue_count = len(original_events) - len(original_dialogue)
        logger.debug(
            f'   📊 Validation: {len(original_dialogue)} dialogue, {non_dialogue_count} non-dialogue (signs/songs/effects)'
        )

        if not original_dialogue:
            logger.warning('No dialogue events found in original file')
            return

        original_start = min(e.start for e in original_dialogue)
        original_end = max(e.end for e in original_dialogue)
        original_duration_sec = (original_end - original_start) / 1000

        first_dialogue = min(original_dialogue, key=lambda e: e.start)
        last_dialogue = max(original_dialogue, key=lambda e: e.end)
        logger.debug(
            f'   📊 Original dialogue range: {original_start}ms - {original_end}ms ({original_duration_sec:.1f}s)'
        )
        logger.debug(f'      First: "{first_dialogue.plaintext.strip()[:50]}..."')
        logger.debug(f'      Last:  "{last_dialogue.plaintext.strip()[:50]}..."')

        if not cleaned_events:
            raise SubtitleProcessingError('Cleaned subtitle file has no events')

        logger.debug(f'   📊 Validation: Cleaned file has {len(cleaned_events)} dialogue events')

        cleaned_start = min(e.start for e in cleaned_events)
        cleaned_end = max(e.end for e in cleaned_events)
        cleaned_duration_sec = (cleaned_end - cleaned_start) / 1000

        first_cleaned = min(cleaned_events, key=lambda e: e.start)
        last_cleaned = max(cleaned_events, key=lambda e: e.end)
        logger.debug(
            f'   📊 Cleaned dialogue range: {cleaned_start}ms - {cleaned_end}ms ({cleaned_duration_sec:.1f}s)'
        )
        logger.debug(f'      First: "{first_cleaned.plaintext.strip()[:50]}..."')
        logger.debug(f'      Last:  "{last_cleaned.plaintext.strip()[:50]}..."')

        TOLERANCE_MS = 50

        start_diff = cleaned_start - original_start
        end_diff = cleaned_end - original_end

        logger.debug('   📊 Timing differences:')
        logger.debug(
            f'      Start: {start_diff:+d}ms ("{"within" if abs(start_diff) <= TOLERANCE_MS else "EXCEEDS"} {TOLERANCE_MS}ms tolerance")'
        )
        logger.debug(
            f'      End:   {end_diff:+d}ms ("{"within" if abs(end_diff) <= TOLERANCE_MS else "EXCEEDS"} {TOLERANCE_MS}ms tolerance")'
        )

        # Timing mismatches are expected when the original has non-dialogue events
        # (signs, effects, "..." placeholders) before/after the dialogue range.
        # The cleaning process filters more aggressively than the style-based filter,
        # so warn but don't fail.
        if abs(start_diff) > TOLERANCE_MS:
            if non_dialogue_count > 0:
                logger.info(
                    f'   ⚠️  Start time offset: {start_diff:+d}ms '
                    f'(expected — original has {non_dialogue_count} non-dialogue events)'
                )
            else:
                logger.warning(f'   ⚠️  Unexpected start time offset: {start_diff:+d}ms')

        if abs(end_diff) > TOLERANCE_MS:
            if non_dialogue_count > 0:
                logger.info(
                    f'   ⚠️  End time offset: {end_diff:+d}ms '
                    f'(expected — original has non-dialogue events after last dialogue)'
                )
            else:
                logger.warning(f'   ⚠️  Unexpected end time offset: {end_diff:+d}ms')

        if abs(start_diff) <= TOLERANCE_MS and abs(end_diff) <= TOLERANCE_MS:
            logger.info('   ✅ Timing validation passed')
        else:
            logger.info('   ✅ Timing validation passed (with expected offsets)')

    @staticmethod
    def _deduplicate_events(subs) -> list:
        """Remove duplicate consecutive events with same text."""
        original_count = len(subs)
        unique_subs = []
        last_text: str | None = None
        current_group_start: int = 0
        current_group_end: int = 0

        pysubs2 = get_pysubs2()
        assert pysubs2 is not None

        for event in subs:
            clean_text = event.plaintext.strip()

            if not clean_text or len(clean_text) < 2:
                continue

            if last_text == clean_text:
                current_group_end = max(current_group_end, event.end)
            else:
                if last_text is not None and unique_subs:
                    consolidated_event = pysubs2.SSAEvent(
                        start=current_group_start,
                        end=current_group_end,
                        style=unique_subs[-1].style if unique_subs else 'Default',
                        text=last_text,
                    )
                    unique_subs[-1] = consolidated_event

                last_text = clean_text
                current_group_start = event.start
                current_group_end = event.end
                unique_subs.append(event)

        if last_text is not None and unique_subs:
            consolidated_event = pysubs2.SSAEvent(
                start=current_group_start,
                end=current_group_end,
                style=unique_subs[-1].style if unique_subs else 'Default',
                text=last_text,
            )
            unique_subs[-1] = consolidated_event

        deduped_count = len(unique_subs)
        if deduped_count < original_count:
            removed = original_count - deduped_count
            logger.debug(
                f'   - Deduplicated: {original_count} → {deduped_count} entries '
                f'(removed {removed} duplicate effect layers)'
            )

        return unique_subs

    @staticmethod
    def _filter_dialogue(events: list) -> list[DialogueLine]:
        """Filter out non-dialogue events and convert to DialogueLine."""
        dialogue_lines = []
        skipped_count = 0

        for event in events:
            if not event.text or event.text.strip() == '':
                skipped_count += 1
                continue

            style = getattr(event, 'style', 'Default').lower()
            if any(keyword in style for keyword in NON_DIALOGUE_STYLES):
                skipped_count += 1
                continue

            clean_text = event.plaintext.strip()
            if not clean_text:
                skipped_count += 1
                continue

            dialogue_lines.append(DialogueLine(event.start, event.end, clean_text))

        logger.info(f'   - Extracted {len(dialogue_lines)} dialogue lines')
        logger.info(f'   - Skipped {skipped_count} non-dialogue events')

        return dialogue_lines

File: crates/mt-subtitles/src/processor.rs
//! Subtitle processing: dialogue extraction, file creation, font overrides,
//! and validation.

use std::path::Path;

use mt_core::types::{DialogueLine, is_non_dialogue_style, replace_polish_chars};
use thiserror::Error;
use tracing::{info, warn};

use crate::{
    ass::{load_ass, to_ass_string},
    error::ParseError,
    model::{Event, EventKind, Subtitles, strip_ass_overrides},
    srt::{load_srt, to_srt_string},
};

/// Error type for subtitle processing failures.
#[derive(Debug, Error)]
pub enum SubtitleProcessingError {
    /// The requested subtitle file does not exist on disk.
    #[error("subtitle file not found: {0}")]
    NotFound(String),

    /// Parsing the subtitle content failed; the underlying [`ParseError`] is
    /// preserved as the error source.
    #[error("failed to parse subtitle file: {0}")]
    Parse(#[from] ParseError),

    /// Writing the subtitle file failed.
    #[error("failed to save subtitle file: {0}")]
    Save(#[source] std::io::Error),

    /// A validation precondition was violated (e.g. cleaned file has no events).
    #[error("{0}")]
    Validation(String),
}

type Result<T> = std::result::Result<T, SubtitleProcessingError>;

/// Find the best dialogue style name in a `Subtitles`.
///
/// Selection logic:
/// 1. If no styles → `"Default"`
/// 2. If `"Default"` exists → `"Default"`
/// 3. Look for `"Dialogue"`, `"Dialog"`, `"Main"`, `"Dialogi"`, `"Normal"` (case-sensitive)
/// 4. Case-insensitive search for any name containing `"dialog"`, `"default"`, or `"main"`
/// 5. Last resort: first style name
pub fn find_dialogue_style(subs: &Subtitles) -> &str {
    if subs.styles.is_empty() {
        return "Default";
    }

    // Prefer "Default" if present
    if subs.styles.iter().any(|s| s.name == "Default") {
        return "Default";
    }

    // Common dialogue names (case-sensitive)
    let preferred = ["Dialogue", "Dialog", "Main", "Dialogi", "Normal"];
    for &name in &preferred {
        if let Some(style) = subs.styles.iter().find(|s| s.name == name) {
            return &style.name;
        }
    }

    // Case-insensitive fallback
    for style in &subs.styles {
        let lower = style.name.to_lowercase();
        if lower.contains("dialog") || lower.contains("default") || lower.contains("main") {
            return &style.name;
        }
    }

    // Last resort: first style
    &subs.styles[0].name
}

/// Unified subtitle processor functions for parsing, writing, and validation.
/// All functions are free-standing (no state needed).
/// Extract dialogue lines from a subtitle file.
pub fn extract_dialogue_lines(subtitle_file: &Path) -> Result<Vec<DialogueLine>> {
    if !subtitle_file.exists() {
        return Err(SubtitleProcessingError::NotFound(
            subtitle_file.display().to_string(),
        ));
    }

    let subs = load_file(subtitle_file)?;

    let unique_events = deduplicate_events(subs.events);
    let dialogue_lines = filter_dialogue(&unique_events);

    Ok(dialogue_lines)
}

/// Create a new subtitle file from dialogue lines.
///
/// Copies `[Script Info]` and `[V4+ Styles]` from `original_file`.
pub fn create_subtitle_file(
    original_file: &Path,
    dialogue_lines: &[DialogueLine],
    output_path: &Path,
    text_transform: Option<fn(&str) -> String>,
) -> Result<()> {
    if !original_file.exists() {
        return Err(SubtitleProcessingError::NotFound(
            original_file.display().to_string(),
        ));
    }

    let mut subs = load_file(original_file)?;

    let dialogue_style = find_dialogue_style(&subs).to_string();

    // Replace events with the provided dialogue lines.
    subs.events = dialogue_lines
        .iter()
        .map(|line| {
            let mut text = line.text.clone();
            if let Some(transform) = text_transform {
                text = transform(&text);
            }
            // \n in text → \N (ASS hard line break)
            text = text.replace('\n', "\\N");
            Event {
                kind: EventKind::Dialogue,
                layer: 0,
                start_ms: line.start_ms,
                end_ms: line.end_ms,
                style: dialogue_style.clone(),
                name: String::new(),
                margin_l: 0,
                margin_r: 0,
                margin_v: 0,
                effect: String::new(),
                text,
            }
        })
        .collect();

    save_file(&subs, output_path)?;
    Ok(())
}

/// Create a clean English subtitle file (no transform).
pub fn create_english_subtitles(
    original_file: &Path,
    dialogue_lines: &[DialogueLine],
    output_path: &Path,
) -> Result<()> {
    create_subtitle_file(original_file, dialogue_lines, output_path, None)
}

/// Create a Polish subtitle file with optional diacritic replacement.
pub fn create_polish_subtitles(
    original_file: &Path,
    translated_dialogue: &[DialogueLine],
    output_path: &Path,
    replace_chars: bool,
) -> Result<()> {
    let transform: Option<fn(&str) -> String> = if replace_chars {
        Some(replace_polish_chars)
    } else {
        None
    };
    create_subtitle_file(original_file, translated_dialogue, output_path, transform)
}

/// Replace all font names in ASS styles with `new_font_name`.
///
/// The style raw field has fontname at index 1 (0-indexed after name).
pub fn override_font_name(ass_file: &Path, new_font_name: &str) -> Result<()> {
    if !ass_file.exists() {
        return Err(SubtitleProcessingError::NotFound(
            ass_file.display().to_string(),
        ));
    }

    let mut subs = load_file(ass_file)?;

    for style in &mut subs.styles {
        let mut owned: Vec<String> = style.raw.split(',').map(|s| s.to_string()).collect();
        // owned[0] = Name, owned[1] = Fontname
        if owned.len() > 1 {
            owned[1] = new_font_name.to_string();
            style.raw = owned.join(",");
        }
    }

    save_file(&subs, ass_file)?;
    Ok(())
}

/// Validate that cleaned subtitles maintain proper timing coverage.
///
/// Timing mismatches are logged as warnings but not fatal.
pub fn validate_cleaned_subtitles(original_file: &Path, cleaned_file: &Path) -> Result<()> {
    let original_subs = load_file(original_file)?;
    let cleaned_subs = load_file(cleaned_file)?;

    let original_events: Vec<_> = original_subs
        .events
        .iter()
        .filter(|e| !e.text.trim().is_empty())
        .collect();
    let cleaned_events: Vec<_> = cleaned_subs
        .events
        .iter()
        .filter(|e| !e.text.trim().is_empty())
        .collect();

    let original_dialogue: Vec<_> = original_events
        .iter()
        .filter(|e| !is_non_dialogue_style(&e.style))
        .copied()
        .collect();
    let non_dialogue_count = original_events.len() - original_dialogue.len();

    if original_dialogue.is_empty() {
        warn!("No dialogue events found in original file");
        return Ok(());
    }

    if cleaned_events.is_empty() {
        return Err(SubtitleProcessingError::Validation(
            "Cleaned subtitle file has no events".to_string(),
        ));
    }

    let original_start = original_dialogue.iter().map(|e| e.start_ms).min().unwrap();
    let original_end = original_dialogue.iter().map(|e| e.end_ms).max().unwrap();
    let cleaned_start = cleaned_events.iter().map(|e| e.start_ms).min().unwrap();
    let cleaned_end = cleaned_events.iter().map(|e| e.end_ms).max().unwrap();

    const TOLERANCE_MS: i64 = 50;
    let start_diff = cleaned_start - original_start;
    let end_diff = cleaned_end - original_end;

    if start_diff.abs() > TOLERANCE_MS {
        if non_dialogue_count > 0 {
            info!(
                start_diff = start_diff,
                non_dialogue_count = non_dialogue_count,
                "start time offset (expected — original has non-dialogue events)"
            );
        } else {
            warn!("unexpected start time offset: {start_diff:+}ms");
        }
    }
    if end_diff.abs() > TOLERANCE_MS {
        if non_dialogue_count > 0 {
            info!(
                end_diff = end_diff,
                non_dialogue_count = non_dialogue_count,
                "end time offset (expected — original has non-dialogue events after last dialogue)"
            );
        } else {
            warn!("unexpected end time offset: {end_diff:+}ms");
        }
    }

    Ok(())
}

/// Remove duplicate consecutive events with the same plain text.
///
/// Consecutive events with identical plain text are merged: the kept event's
/// `end_ms` = max of group `end_ms`, and its text is normalized to the stripped
/// plain text. Events with plain text shorter than 2 characters are dropped.
pub fn deduplicate_events(events: Vec<Event>) -> Vec<Event> {
    let mut unique: Vec<Event> = Vec::new();
    let mut last_text: Option<String> = None;
    let mut group_end: i64 = 0;

    for mut event in events {
        let clean_text = strip_ass_overrides(&event.text).trim().to_string();

        if clean_text.is_empty() || clean_text.chars().count() < 2 {
            continue;
        }

        if last_text.as_deref() == Some(clean_text.as_str()) {
            group_end = group_end.max(event.end_ms);
            if let Some(last) = unique.last_mut() {
                last.end_ms = group_end;
            }
        } else {
            // Normalize the kept event's text to its stripped plain text.
            event.text = clean_text.clone();
            last_text = Some(clean_text);
            group_end = event.end_ms;
            unique.push(event);
        }
    }

    unique
}

/// Filter events to dialogue only, converting to `DialogueLine`.
///
/// Skips events with:
/// - empty text
/// - style matching any `NON_DIALOGUE_STYLES` keyword (case-insensitive)
/// - plain text (after stripping override tags) that is empty
fn filter_dialogue(events: &[Event]) -> Vec<DialogueLine> {
    let mut result = Vec::new();

    for event in events {
        if event.text.trim().is_empty() {
            continue;
        }

        if is_non_dialogue_style(&event.style) {
            continue;
        }

        // Match pysubs2 `SSAEvent.plaintext`: strip override tags, then
        // convert ASS hard breaks (`\N`) and soft breaks (`\n`) into real
        // newlines so translation models do not receive literal `\N` tokens.
        let clean = strip_ass_overrides(&event.text)
            .replace("\\N", "\n")
            .replace("\\n", "\n")
            .trim()
            .to_string();
        if clean.is_empty() {
            continue;
        }

        result.push(DialogueLine {
            start_ms: event.start_ms,
            end_ms: event.end_ms,
            text: clean,
        });
    }

    result
}

fn load_file(path: &Path) -> Result<Subtitles> {
    let content = std::fs::read_to_string(path).map_err(ParseError::Io)?;
    match path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_ascii_lowercase())
        .as_deref()
    {
        Some("ass") | Some("ssa") => Ok(load_ass(&content)?),
        Some("srt") => Ok(load_srt(&content)?),
        other => Err(ParseError::UnsupportedExtension(other.map(String::from)).into()),
    }
}

fn save_file(subs: &Subtitles, path: &Path) -> Result<()> {
    let content = match path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_ascii_lowercase())
        .as_deref()
    {
        Some("srt") => to_srt_string(subs),
        _ => to_ass_string(subs),
    };
    std::fs::write(path, content).map_err(SubtitleProcessingError::Save)
}
